C3 s 511 v2 mpqb include hist res

BenMGeo opened the pull request:
@bascrezee 
I try to use the branch, but can't get it to run.

There is `"mask_glaciated"` missing ( I updated from the development branch, but no) and somehow I cannot make the taylordiagramm executable (though I used the noted command: `chmod -x [...]/esmvaltool/diag_scripts/mpqb/taylordiagram.py`).


So implementation has to wait till Monday ;(

bascrezee commented:
> @bascrezee
> I try to use the branch, but can't get it to run.
> 
> There is `"mask_glaciated"` missing ( I updated from the development branch, but no) 
I implemented some new functionality, please try the branch `https://github.com/ESMValGroup/ESMValCore/tree/mask_glaciated`



and somehow I cannot make the taylordiagramm executable (though I used the noted command: `chmod -x [...]/esmvaltool/diag_scripts/mpqb/taylordiagram.py`).
> 

Should be `chmod +x`, right?

BenMGeo commented:
> > @bascrezee
> > I try to use the branch, but can't get it to run.
> > There is `"mask_glaciated"` missing ( I updated from the development branch, but no)
> 
> I implemented some new functionality, please try the branch `https://github.com/ESMValGroup/ESMValCore/tree/mask_glaciated`

I try to join this in. Should work without that preprocessing anyways. So for testing purposes, it should be fine.

> 
> > and somehow I cannot make the taylordiagramm executable (though I used the noted command: `chmod -x [...]/esmvaltool/diag_scripts/mpqb/taylordiagram.py`).
> 
> Should be `chmod +x`, right?

yes and it is missing the python headline `#!/usr/bin/env python`.

Now I end up with the following error:

```
Traceback (most recent call last):
  File "/media/bmueller/Work/GIT/ESMValTool_V2_c3s511/esmvaltool/diag_scripts/mpqb/taylordiagram.py", line 8, in <module>
    import skill_metrics
ModuleNotFoundError: No module named 'skill_metrics'
```

bascrezee commented:
> > > @bascrezee
> > > I try to use the branch, but can't get it to run.
> > > There is `"mask_glaciated"` missing ( I updated from the development branch, but no)
> > 
> > 
> > I implemented some new functionality, please try the branch `https://github.com/ESMValGroup/ESMValCore/tree/mask_glaciated`
> 
> I try to join this in. Should work without that preprocessing anyways. So for testing purposes, it should be fine.
> 
> > > and somehow I cannot make the taylordiagramm executable (though I used the noted command: `chmod -x [...]/esmvaltool/diag_scripts/mpqb/taylordiagram.py`).
> > 
> > 
> > Should be `chmod +x`, right?
> 
> yes and it is missing the python headline `#!/usr/bin/env python`.
> 
> Now I end up with the following error:
> 
> ```
> Traceback (most recent call last):
>   File "/media/bmueller/Work/GIT/ESMValTool_V2_c3s511/esmvaltool/diag_scripts/mpqb/taylordiagram.py", line 8, in <module>
>     import skill_metrics
> ModuleNotFoundError: No module named 'skill_metrics'
> ```

-->  `pip install SkillMetrics`

BenMGeo commented:
Added the package to setup.py.

Now I end up with:
```
Traceback (most recent call last):
  File "/media/bmueller/Work/GIT/ESMValTool_V2_c3s511/esmvaltool/diag_scripts/mpqb/taylordiagram.py", line 15, in <module>
    from diag1d import rmsd1d
ModuleNotFoundError: No module named 'diag1d'
```

Is this the one in your trend analysis? So I'd put this into a meta dictionary that can be accessed from multiple codes?

bascrezee commented:
> Added the package to setup.py.
> 
> Now I end up with:
> 
> ```
> Traceback (most recent call last):
>   File "/media/bmueller/Work/GIT/ESMValTool_V2_c3s511/esmvaltool/diag_scripts/mpqb/taylordiagram.py", line 15, in <module>
>     from diag1d import rmsd1d
> ModuleNotFoundError: No module named 'diag1d'
> ```
> 
> Is this the one in your trend analysis? So I'd put this into a meta dictionary that can be accessed from multiple codes?

Yes, exactly, it is that one :)

BenMGeo commented:
So do you want to share this as trend framework or do you want to share the minimum overlap?Do you want the trend framework to be installed with the tool?

Your code, you decide.

bascrezee commented:
> So do you want to share this as trend framework or do you want to share the minimum overlap?Do you want the trend framework to be installed with the tool?
> 
> Your code, you decide.

Ideally, I wanted to get the trend framework fully into the ESMValTool. But I think this is not the right moment, there are several PRs still open, that have priority over inclusion of this trend framework. However, in the end, it would be nice to have it in, including a proper review of all the code.

For now, please use your local copy of the trend framework which can be accessed from ESMValTool if is included in your $PYTHONPATH.

BenMGeo commented:
another preprocessor issue:

```
ValueError: Invalid argument(s): standardize encountered for preprocessor function anomalies. 
Valid arguments are: [period]
```

This functionality is not in the development version.

bascrezee commented:
Are you using  the right branch of ESMValCore? see above, both features
should be in there.

On Tue, 3 Dec 2019, 07:14 Benjamin Müller, <notifications@github.com> wrote:

> another preprocessor issue:
>
> ValueError: Invalid argument(s): standardize encountered for preprocessor function anomalies.
> Valid arguments are: [period]
>
> This functionality is not in the development version.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMValGroup/ESMValTool/pull/1456?email_source=notifications&email_token=AGTSE5A53LEAKWSRD3VO77LQWX2NPA5CNFSM4JTA4SZ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFYHJZQ#issuecomment-561018086>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGTSE5GMRRGB7JXAOQJOAPDQWX2NPANCNFSM4JTA4SZQ>
> .
>

BenMGeo commented:
I updated the tool not the core... my bad.

BenMGeo commented on esmvaltool/diag_scripts/mpqb/mpqb_auxmaps.py:
```diff
@@ -6,11 +6,11 @@
 import numpy as np
 
-from diag1d import mannkendall1d, theilslopes1d
+from esmvaltool.diag_scripts.shared.trend_mpqb_common.diag1d import mannkendall1d, theilslopes1d
```
@bascrezee 
I added the functions missing in this branch to the shared directory, so we can refer to them in the trend package as well.

BenMGeo commented on setup.py:
```diff
@@ -44,6 +44,7 @@
         'xarray>=0.12',
         'xlrd',
         'xlsxwriter',
+        'SkillMetrics',
```
I included the SkillMetrrics package to the installation routine.

BenMGeo commented on esmvaltool/diag_scripts/mpqb/taylordiagram.py:
```diff
@@ -24,11 +24,9 @@
-        reference_dataset = recipe_cfg['reference_dataset']
-
+    # read referenece data set
+    reference_dataset = cfg['reference_dataset']
```
This moves the information from a separate config file to the main recipe

BenMGeo commented on esmvaltool/recipes/mpqb/recipe_mpqb_test_BM.yml:
```diff
@@ -21,6 +21,10 @@ timeperiod: &timeperiod
 #  - {dataset: CDS-UERRA, type: reanaly, project: OBS6, mip: E6hr,
 #     version: UERRA-HARMONIE, tier: 3, frequency: 6hr, <<: *timeperiod}
 
+mmm_stat: &mmm_stat
```
Adds the multimodel-mean and -median statistic to the preprocessor (this is automatically added to the datasets).

BenMGeo commented on esmvaltool/recipes/mpqb/recipe_mpqb_test_BM.yml:
```diff
@@ -38,6 +42,7 @@ default_pp_maps: &pp_maps
   <<: *mpqbregridder
   mask_landsea:
     mask_out: sea
+  <<: *mmm_stat
```
add the multimodel to the standard regridder.

BenMGeo commented on esmvaltool/diag_scripts/mpqb/mpqb_comparison.py:
```diff
@@ -130,10 +132,8 @@ def compute_diagnostic(filename):
-        recipe_cfg = yaml.safe_load(handle)
-        reference_dataset = recipe_cfg['reference_dataset']
+    # read referenece data set
+    reference_dataset = cfg['reference_dataset']
```
Here also reads reference dataset from recipe instead of config.

BenMGeo commented on esmvaltool/recipes/mpqb/recipe_mpqb_test_BM.yml:
```diff
@@ -135,119 +135,107 @@ preprocessors:
+  exp: historical
+  start_year: 1990
+  end_year: 2000
 
```
Grouping these makes it easier to read

BenMGeo commented:
I also updated the basic tool from the version2 development branch, so it is easier to check missing functionalities. 
Maybe we have to cherry-pick, though, for the final PR.

bascrezee commented:
> I also updated the basic tool from the version2 development branch, so it is easier to check missing functionalities.
> Maybe we have to cherry-pick, though, for the final PR.

I think at the moment this PR is way too large, there seem to be a lot of files coming along with your PR that should not be there (e.g. a .gitignore script that is different, many NCL scripts). It would be nice to keep the branch clean, having only the MPQB diagnostics such that it can be easily merged into version2_development at a later stage.

BenMGeo commented:
Yes, I'll clean the branch and prepare a cleaner one once I added missing code. 

I updated from the development branch, this is where all the differences came from, to add the functionality I was missing.

BenMGeo commented:
Example for histogram: (see recipe)

![histogram](https://user-images.githubusercontent.com/14977178/70152262-2ab6d580-16ad-11ea-95a0-025de77fc7c6.png)

bascrezee commented:
Very nice! I did like the opaque histograms as well actually. Maybe we can include both?

BenMGeo commented:
I realized that, with more than 5 histograms, colors are lost. This way, we also keep the same color for both histograms and line plots.

bascrezee commented:
> I realized that, with more than 5 datasets, colors are lost. This way, we also keep the same color for both histograms and line plots.

We can still do that, but with a different type of histogram. Using `histtype='step'` should produce something like this: 

https://i.stack.imgur.com/FemR3.png (just a random plot from the internet, showing what it looks like)

BenMGeo commented:
Not that easy in the current realization. If you want to have lazy histograms you have to calculate the height and bins first and then plot them as bars.

I'm not sure if bars has a similar functionality.

bascrezee commented:
> Not that easy in the current realization. If you want to have lazy histograms you have to calculate the height and bins first and then plot them as bars.
> 
> I'm not sure if bars has a similar functionality.

Ok, I thought it would be straightforward to implement, then leave it for now. I might give it a try later on.

bascrezee commented:
> Not that easy in the current realization. If you want to have lazy histograms you have to calculate the height and bins first and then plot them as bars.
> I'm not sure if bars has a similar functionality.
 

Ok, I thought it would be straightforward to implement, then leave it for now. I might give it a try later on.

BenMGeo commented:
Example for resolution plot: (see recipe)
![resolution](https://user-images.githubusercontent.com/14977178/70226351-014f8580-1751-11ea-9926-fd61fe9d1782.png)

BenMGeo commented:
I would clean this up now and re-request a draft pull

bascrezee commented:
@BenMGeo Just for info. I found out what is creating the grey shading east of the meridian, I have the same problem, working on it.
